src/useForm.ts

```diff
@@ -0,0 +1,489 @@
+
+      return () => {
+        names.forEach((name) => fields.current[name].callbacks.delete(callback));
+      };
+    };
+
+    const getOnChange =
+      <N extends Name>(name: N) =>
+      (value: Values[N]): void => {
+        setState(name, (prevState) => ({
+          ...prevState,
+          value,
+        }));
+
+        setTalkative(name, ["onChange"]);
+
+        if (formStatus.current === "untouched" || formStatus.current === "submitted") {
+          formStatus.current = "editing";
+          forceUpdate();
+        }
+
+        void internalValidateField(name);
+      };
+
+    const getOnBlur = (name: Name) => (): void => {
+      const { validity } = fields.current[name].state;
+
+      // Avoid validating an untouched / already valid field
+      if (validity.tag !== "unknown" && !isTalkative(name)) {
+        setTalkative(name, ["onBlur", "onSuccessOrBlur"]);
+        void internalValidateField(name);
+      }
+    };
+
+    const resetForm: Contract["resetForm"] = () => {
+      Object.keys(arg.current).forEach((name) => resetField(name));
+      formStatus.current = "untouched";
+
+      forceUpdate();
+    };
+
+    const focusFirstError = (names: Name[], results: ValidatorResult<ErrorMessage>[]) => {
+      const index = results.findIndex((result) => typeof result !== "undefined");
+      const name = names[index];
+
+      if (typeof name !== "undefined") {
+        focusField(name);
+      }
+    };
+
+    const isSuccessfulSubmission = (
+      results: ValidatorResult<ErrorMessage>[],
+    ): results is undefined[] => results.every((result) => typeof result === "undefined");
+
+    const submitForm: Contract["submitForm"] = ({
+      onSuccess = noop,
+      onFailure = noop,
+      focusOnFirstError = true,
+    } = {}) => {
+      if (formStatus.current === "submitting") {
+        return; // Avoid concurrent submissions
+      }
+
+      formStatus.current = "submitting";
+
+      const keys: Name[] = Object.keys(fields.current);
+      const names = keys.filter((name) => fields.current[name].mounted);
+      const values = {} as OptionalRecord<Values>;
+      const errors: Partial<Record<Name, ErrorMessage>> = {};
+      const results: ValidatorResult<ErrorMessage>[] = [];
+
+      keys.forEach((name) => {
+        values[name] = Option.None();
+      });
+
+      names.forEach((name: Name, index) => {
+        setTalkative(name);
+        values[name] = Option.Some(getFieldValue(name, { sanitize: true }));
+        results[index] = internalValidateField(name);
+      });
+
+      if (isSuccessfulSubmission(results)) {
+        const effect = onSuccess(values);
+
+        if (!isPromise(effect)) {
+          formStatus.current = "submitted";
+          return forceUpdate();
+        }
+
+        forceUpdate();
+
+        void effect.finally(() => {
+          formStatus.current = "submitted";
+
+          if (mounted.current) {
+            forceUpdate();
+          }
+        });
```

Comment:
could we replace it with `Promise.resolve(effect)` and have a single path?

Reply:
Not really. If you don't return a `promise`, there's no need for an intermediate re-render with `formStatus` = `submitting` as the form submission appears to be sync.

Reply:
(In case the effect is not a promise, there's no `forceUpdate()` between `formStatus.current = "submitting"` and `formStatus.current = "submitted"`)

Reply:
in that case we can have:

```ts
const nextTick = Promise.prototype.then.bind(Promise.resolve());
```

and in `submitForm`:

```ts
nextTick(() => {
  formStatus.current = "submitting";
  forceUpdate();
});

effect.then(() => {
  formStatus.current = "submitted";
  forceUpdate();
});
```

which eliminates the double render-phase

Reply:
We can also use:

```ts
const setFormSubmitted = () => {
  formStatus.current = "submitted";

  if (mounted.current) {
    forceUpdate();
  }
};
```

```ts
if (isSuccessfulSubmission(results)) {
  const effect = onSuccess(values);

  if (isPromise(effect)) {
    forceUpdate();
    void effect.finally(setFormSubmitted);
  } else {
    setFormSubmitted();
  }
}
```

It's easy to understand